gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+            "JPEGXR - Decoder initialization failed : %s", e.what());
+        return nullptr;
+    }
+
+/* -------------------------------------------------------------------- */
+/*      Create band information objects.                                */
+/* -------------------------------------------------------------------- */
+    CPLAssert(nBands > 0);
+
+    for (int iBand = 1; iBand <= nBands; iBand++)
+    {
+        poDS->SetBand(iBand, new JPEGXRRasterBand(poDS, iBand));
+    }
+
+    CPLAssert(poDS->nBands == nBands);
+
+/* -------------------------------------------------------------------- */
+/*      Initialize any PAM information.                                 */
+/* -------------------------------------------------------------------- */
+    poDS->SetDescription(poOpenInfo->pszFilename);
+    poDS->TryLoadXML();
+
+    return poDS;
+}
+
+/************************************************************************/
+/*                             Uncompress()                             */
+/************************************************************************/
+
+CPLErr JPEGXRDataset::Uncompress()
+{
+    try
+    {
+        if (bHasUncompressed)
+            return CE_None;
+
+        size_t nUncompressedSize = nRasterXSize * nRasterYSize * nBands
+            * GDALGetDataTypeSizeBytes(GetRasterBand(1)->GetRasterDataType());
+        pabyUncompressedData = (GByte*)VSI_MALLOC_VERBOSE(nUncompressedSize);
+        if (!pabyUncompressedData)
+            return CE_Failure;
+
+        oDecoder.Read(pabyUncompressedData, 0, 0, nRasterXSize, nRasterYSize);
+
+        bHasUncompressed = true;
+        return CE_None;
+    }
+    catch (JPEGXRException const& e)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+            "JPEGXR - Decompression of data failed : %s", e.what());
+    }
+    return CE_Failure;
+}
+
+/************************************************************************/
+/*                          GDALRegister_JPEGXR()                       */
+/************************************************************************/
+
+void GDALRegister_JPEGXR()
+
+{
+    if(GDALGetDriverByName("JPEGXR") != nullptr)
+        return;
+
+    GDALDriver* poDriver = new GDALDriver();
+
+    poDriver->SetDescription( "JPEGXR" );
+    poDriver->SetMetadataItem( GDAL_DCAP_RASTER, "YES" );
+    poDriver->SetMetadataItem( GDAL_DMD_LONGNAME,
+                               "JPEG XR driver based on jxrlib library" );
+    poDriver->SetMetadataItem( GDAL_DMD_HELPTOPIC, "frmt_jpegxr.html" );
+    // NOTE: The HD Photo format is a pre-standard implementation of the JPEG XR format.
+    // https://msdn.microsoft.com/en-us/library/windows/desktop/gg430023.aspx
+    poDriver->SetMetadataItem( GDAL_DMD_EXTENSIONS, ".jxr .hdp .wdp" );
+    poDriver->SetMetadataItem( GDAL_DMD_MIMETYPE, "image/vnd.ms-photo" );
+    poDriver->SetMetadataItem( GDAL_DMD_CREATIONDATATYPES, "Byte" ); // TODO
+
```

Comment:
Setting GDAL_DMD_CREATIONOPTIONLIST to advertize the creation options would be good.